Libraries/DirectMLX.h

```diff
@@ -3422,9 +3422,64 @@ namespace dml
     // TODO: Argmax
     // 
 
-    // 
-    // TODO: ROIAlign
-    // 
+#if DML_TARGET_VERSION >= 0x4000
+
+    inline Expression RoiAlign(
+        Expression input,
+        Expression roi,
+        Expression batchIndices,
+        DML_REDUCE_FUNCTION reductionFunction,
+        DML_INTERPOLATION_MODE interpolationMode,
+        float spatialScaleX,
+        float spatialScaleY,
+        float inputPixelOffset,
+        float outputPixelOffset,
+        float outOfBoundsInputValue,
+        uint32_t minimumSamplesPerOutput,
+        uint32_t maximumSamplesPerOutput,
+        bool alignRegionsToCorners,
+        uint32_t cropHeight,
+        uint32_t cropWidth)
+    {
+        detail::GraphBuilder* builder = input.Impl()->GetGraphBuilder();
+
+        TensorDesc inputTensor = input.Impl()->GetOutputDesc();
+        TensorDesc roiTensor = roi.Impl()->GetOutputDesc();
+        TensorDesc batchIndicesTensor = batchIndices.Impl()->GetOutputDesc();
+
+        TensorDesc::Dimensions outputSizes({
+            roiTensor.sizes[roiTensor.sizes.size() - 2],
```

Comment:
The ROI can accept dimension count 2 to 4.

```
        TensorDescParameter roiTensor =
        {
            "ROI",
            desc.ROITensor,
            ParameterType::RequiredInput,
            TensorDataTypeMask::Float16to32,
            2,4,                    // min/max rank
            TensorIndex::Roi,       // self index
            TensorIndex::Input,     // matches type
            TensorIndex::None,      // matches rank
            TensorIndex::None,      // matches sizes
        };
```

Where any leading dimensions are treated as 1's. So I'd split this out into:

```
uint32_t roiCount = roiTensor.sizes.size() < 2 ? 1 : roiTensor.sizes[roiTensor.sizes.size() - 2];
 TensorDesc::Dimensions outputSizes({
    roiCount,
    channelCount,
    outputHeight,
    outputWidth // not "crop height", which is confusing and comes from TF terminology, because the crop size actually varies per individual ROI
});
```